### PR TITLE
Electrical system V0.91 V/IR  (still needs cleaning and spite work)

### PR DESCRIPTION
**Pull request by Bod9001**
### Purpose
Electrical system
### Open Questions and Pre-Merge TODOs

- [ ]  This fix is tested on the same branch it is PR'ed to.  Since I've been working on it for a long time it's a bit out of date so I tried  Compatible with merging but It might not work... **Crosses fingers hoping it works**
- [not yet ]  I correctly commented my code
- [x]  My code is indented with tabs and not spaces
- [x]  This PR does not include any unnecessary .meta, .prefab or .unity (scene) changes
- [x]  This PR does not bring up any new compile errors
- [x]  This PR has been tested in editor
- [ Something I need to do ]  This PR has been tested in multiplayer

### Notes:
ElectricalManager (needed for the electrical system to work) Needs to be added to the managers prefab Doing it on This end causes conflicts

### CODE REVIEW

**Comment by DooblyNoobly** [comment]
@Bod9001 watch the foreach use. Remember to use for loops if you can as it reduces the risk of creating too much garbage. Also there is a fair bit of white space and poor formatting. Will merge this PR but try to clean up the scripts a little bit on the next PR.

**Comment by DooblyNoobly** [comment]
Ah just noticed your note about needing cleaning :) all good

File: UnityProject/Assets/Scripts/Electricity/PowerSupply/PowerSupply.cs [hunk 1]
```diff
@@ -4,159 +4,234 @@
...
+	public IElectricalNeedUpdate RelatedUpdateDevice {get; set;}
+
+	[Header("0 = conn to wire on same tile")]
+	public int DirectionStart;
+	public int DirectionEnd;
+
+	public HashSet<IElectricityIO> ResistanceToConnectedDevices {get; set;} = new HashSet<IElectricityIO>();
+	public HashSet<IElectricityIO> connectedDevices {get; set;} = new HashSet<IElectricityIO>();
+	public List<IElectricityIO> connections {get; set;} = new List<IElectricityIO> ();
+	public Dictionary<int,Dictionary<IElectricityIO,float>> CurrentGoingTo{get; set;} = new Dictionary<int, Dictionary<IElectricityIO, float>> ();
```
Comment (fomalsd) [inline UnityProject/Assets/Scripts/Electricity/PowerSupply/PowerSupply.cs:19]: You might be better off extracting these many common fields to a common abstract class, as things like PoweredDevice, InLineDevice and several more have a really similar set of fields. You'll have them in one place and not have to worry about making a typo in one of classes or changing type and whatnot

**Comment by fomalsd** [comment]
I'm not too sure, but it looks like the entire electricity simulation code runs for each client, too?
It's supposed to run on serverside only and just update sprites/respond to multitool inquiries from clients when asked, from what I imagine

**Comment by Bod9001** [comment]
@fomalsd yeah, I don't have much knowledge of networking, so I will need to ask when I start working on that side

File: UnityProject/Assets/Scripts/Electricity/PowerSupply/PowerSupply.cs (hunk 1 above)
Comment (Bod9001) [reply UnityProject/Assets/Scripts/Electricity/PowerSupply/PowerSupply.cs:19]: yeah, I will get on that

**Pull request merged**